src/repair.rs

```diff
@@ -93,164 +100,223 @@ impl<N: Network> Repair<N> {
+                // handle request for repairing new block
+                Some(block_id) = repair_receiver.recv() => {
+                    self.repair_block(block_id).await;
+                }
+                // handle next request timeout
+                _ = tokio::time::sleep(sleep_duration) => {
+                    let Some((_, hash)) = self.request_timeouts.pop() else {
+                        continue;
+                    };
+                    if let Some(request) = self.outstanding_requests.remove(&hash) {
+                        debug!("retrying timed-out repair request {request:?}");
+                        self.send_request(request).await.unwrap();
+                    }
+                }
+            }
         }
     }
 
     /// Starts repair process for the block specified by `slot` and `block_hash`.
-    pub async fn repair_block(&self, slot: Slot, block_hash: Hash) {
-        if self
-            .blockstore
-            .read()
-            .await
-            .get_block(slot, block_hash)
-            .is_some()
-        {
+    pub async fn repair_block(&mut self, block_id: BlockId) {
+        let (slot, block_hash) = block_id;
+        let h = &hex::encode(block_hash)[..8];
+        if self.blockstore.read().await.get_block(block_id).is_some() {
+            trace!("ignoring repair for block {h} in slot {slot}, already have the block");
             return;
         }
 
-        let h = &hex::encode(block_hash)[..8];
         debug!("repairing block {h} in slot {slot}");
-        // TODO: perform actual repair
-        // HACK: magic number of 10 requests (to ensure it can handle some failures)
-        for _ in 0..10 {
-            self.request_parent(slot, block_hash).await.unwrap();
+        let req = RepairRequest::LastSliceRoot(block_id);
+        self.send_request(req).await.unwrap();
+    }
+
+    async fn handle_repair_message(&mut self, msg: RepairMessage) -> Result<(), NetworkError> {
+        match msg {
+            RepairMessage::Request(request, sender) => {
+                self.answer_request(request, sender).await?;
+            }
+            RepairMessage::Response(resposne) => {
+                self.handle_response(resposne).await;
+            }
         }
+        Ok(())
     }
 
     /// Tries to answer the given repair request.
     ///
     /// If we do not have the block the request refers to, the request is ignored.
     /// Otherwise, the correct response is sent back to the sender of the request.
-    pub async fn answer_request(&self, request: RepairRequest) -> Result<(), NetworkError> {
+    async fn answer_request(
+        &self,
+        request: RepairRequest,
+        sender: ValidatorId,
+    ) -> Result<(), NetworkError> {
         trace!("answering repair request: {request:?}");
-        let slot = request.slot();
-        let hash = request.block_hash();
+        let block_id = request.block_id();
```

Comment:
nit: I would probably not use this and look up the block_id explicitly in the match below.  

Having code structured like this causes some pain for the reader:

- when I look at the individual request handling and see that it contains info that is ignored, I wonder why was that info included in the first place.
- Makes me wonder if I have to look into the implementation of `request.block_id()` to see if something extra has to be done to figure the block id out.